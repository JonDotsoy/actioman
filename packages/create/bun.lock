{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "create-actioman",
      "dependencies": {
        "@jondotsoy/flags": "^2.3.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@jondotsoy/console-draw": ["@jondotsoy/console-draw@0.1.2", "", { "dependencies": { "@jondotsoy/style-text": "^0.9.1" }, "peerDependencies": { "prettier": "^3.3.3", "typescript": "^5.0.0" } }, "sha512-BBZGLd6RzjKxKcgD2OwD7v6GeKlf53Ysk59H+7+HvyGLcJZ7AJk8rOCq/yv813KbWV2teTCEBQXnnP5hZe/2CA=="],

    "@jondotsoy/flags": ["@jondotsoy/flags@2.3.0", "", { "dependencies": { "@jondotsoy/console-draw": "^0.1.2" } }, "sha512-vUMe4NdXKv8RxMXrKCJLsHBt3XQ8/YV4Fyr1I43vSLe5v0M8sdY3wJl1qdTgwRBZES2zMLoenxOUSNGE9r/5NA=="],

    "@jondotsoy/style-text": ["@jondotsoy/style-text@0.9.2", "", { "peerDependencies": { "typescript": "^5.0.0" } }, "sha512-6fonOzrANwrw+UIM4fivtDZLvuCN00yH0lCEYi14ln4j8fSHZokb6TLrqOu38GCqqQnZ2LuzJTQBWgv7jN6iVA=="],

    "@types/bun": ["@types/bun@1.2.2", "", { "dependencies": { "bun-types": "1.2.2" } }, "sha512-tr74gdku+AEDN5ergNiBnplr7hpDp3V1h7fqI2GcR/rsUaM39jpSeKH0TFibRvU0KwniRx5POgaYnaXbk0hU+w=="],

    "@types/node": ["@types/node@22.13.1", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-jK8uzQlrvXqEU91UxiK5J7pKHyzgnI1Qnl0QDHIgVGuolJhRb9EEl28Cj9b3rGR8B2lhFCtvIm5os8lFnO/1Ew=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.2", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-RCbMH5elr9gjgDGDhkTTugA21XtJAy/9jkKe/G3WR2q17VPGhcquf9Sir6uay9iW+7P/BV0CAHA1XlHXMAVKHg=="],

    "prettier": ["prettier@3.4.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-e9MewbtFo+Fevyuxn/4rrcDAaq0IYxPGLvObpQjiZBMAzB9IGmzlnG9RZy3FFas+eBMu2vA0CszMeduow5dIuQ=="],

    "typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
