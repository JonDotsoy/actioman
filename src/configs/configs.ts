import { get } from "@jondotsoy/utils-js/get";
import type { Middleware } from "artur/http/router";

export type IntegrationModule = Integration;

export class Integration {
  constructor(
    readonly name: string,
    readonly hooks?: {
      "http:middleware"?: Middleware<any> | undefined;
    },
  ) {}

  static fromModule(module: unknown): Integration | null {
    const name = get.string(module, "name");
    if (!name) return null;
    const hooks = get.object(module, "hooks");

    const httpMiddleware: any = get.function(hooks, "http:middleware");

    return new Integration(name, {
      "http:middleware": httpMiddleware,
    });
  }
}

const stringBooleanMap: Record<string, boolean | undefined> = {
  true: true,
  "1": true,
  on: true,
  yes: true,
  false: false,
  "0": false,
  off: false,
  no: false,
};

const getStringBoolean = (obj: unknown, ...path: string[]) => {
  const value = get(obj, ...path);
  if (typeof value === "boolean") return value;
  if (typeof value === "string") return stringBooleanMap[value.toLowerCase()];
  return undefined;
};

export type ConfigsModule = {
  integrations?: IntegrationModule[];
};

export class Configs {
  integrations?: Integration[] = [];

  private constructor() {}

  getHTTPListenerMiddlewares(): Middleware<any>[] {
    return (
      this.integrations
        ?.map((integration) => integration.hooks?.["http:middleware"])
        .filter((middleware) => middleware !== undefined) ?? []
    );
  }

  static fromModule(module: unknown) {
    const configs = new Configs();

    const integrations: Integration[] = [];

    for (const integrationModule of get.array(module, "integrations") ?? []) {
      const integration = Integration.fromModule(integrationModule);
      if (!integration) continue;
      integrations.push(integration);
    }

    configs.integrations = integrations;

    return configs;
  }
}
