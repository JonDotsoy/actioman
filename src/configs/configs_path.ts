export const CONFIG_PATH = new URL("./configs.js", import.meta.url);
