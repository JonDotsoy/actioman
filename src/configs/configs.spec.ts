import { describe, expect, it } from "bun:test";
import { Configs } from "./configs";

describe("Configs", () => {
  it.todo("loads YAML configs", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory(
        new URL("__samples__/configs.config.yaml", import.meta.url).toString(),
      ),
    ).toMatchSnapshot();
  });
  it.todo("loads JSON configs", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory(
        new URL("__samples__/configs.config.json", import.meta.url).toString(),
      ),
    ).toMatchSnapshot();
  });
  it.todo("loads TOML configs", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory(
        new URL("__samples__/configs.config.toml", import.meta.url).toString(),
      ),
    ).toMatchSnapshot();
  });
  it("loads JS configs", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory(
        new URL("__samples__/configs.config.js", import.meta.url).toString(),
      ),
    ).toMatchSnapshot();
  });
  it("loads TS configs", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory(
        new URL("__samples__/configs.config.ts", import.meta.url).toString(),
      ),
    ).toMatchSnapshot();
  });
  it("loads MJS configs", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory(
        new URL("__samples__/configs.config.mjs", import.meta.url).toString(),
      ),
    ).toMatchSnapshot();
  });
  it("loads configs from the first file found", async () => {
    const { default: factory } = await import("./modules/factory.js");

    expect(
      await factory.findOn(
        new URL("__samples__/", import.meta.url).toString(),
        "configs",
      ),
    ).toMatchSnapshot();
  });
});

describe("Configs.fromModule", () => {
  it("returns a Configs instance", () => {
    const configs = Configs.fromModule({});

    expect(configs).toBeInstanceOf(Configs);
  });

  it("loads configs from a module", async () => {
    const { default: factory } = await import("./modules/factory.js");

    const configs = Configs.fromModule(
      await factory.findOn(
        new URL("__samples__/", import.meta.url).toString(),
        "configs",
      ),
    );

    expect(configs).toBeInstanceOf(Configs);
  });
});
