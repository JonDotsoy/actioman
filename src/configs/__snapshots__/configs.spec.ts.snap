// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Configs loads JS configs 1`] = `
{
  "hello": true,
}
`;

exports[`Configs loads TS configs 1`] = `
{
  "hello": true,
}
`;

exports[`Configs loads MJS configs 1`] = `
{
  "hello": true,
}
`;

exports[`Configs loads configs from the first file found 1`] = `
{
  "hello": true,
}
`;
